File: 3rd Sem/NM/python/matrixfactorization.py
def lu_decomposition(a, n):
    L = [[0] * n for _ in range(n)]  # Initialize lower triangular matrix L
    U = [[0] * n for _ in range(n)]  # Initialize upper triangular matrix U
    
    # Decompose the matrix 'a' into L and U
    for i in range(n):
        # Set diagonal of L to 1
        L[i][i] = 1
        
        # Upper triangular matrix U
        for j in range(i, n):
            U[i][j] = a[i][j] - sum(L[i][k] * U[k][j] for k in range(i))
        
        # Lower triangular matrix L
        for j in range(i + 1, n):
            L[j][i] = (a[j][i] - sum(L[j][k] * U[k][i] for k in range(i))) / U[i][i]
    
    return L, U

# Input: Matrix size
n = int(input("Enter matrix size: "))

# Input: Matrix 'a' as a list of lists (each row inputted separately)
a = [list(map(float, input(f"Enter row {i+1}: ").split())) for i in range(n)]

# Call LU decomposition
L, U = lu_decomposition(a, n)

# Output: Print L and U matrices
print("L (Lower Triangular Matrix):")
for row in L:
    print(*row)

print("U (Upper Triangular Matrix):")
for row in U:
    print(*row)

File: 3rd Sem/NM/C-program/matrixfactorization.c
#include <stdio.h>
#include <stdlib.h>

void lu_decomposition(int n, double **a, double **L, double **U) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            L[i][j] = 0;
            U[i][j] = 0;
        }
        L[i][i] = 1;
    }

    for (int i = 0; i < n; i++) {
        for (int j = i; j < n; j++) {
            double sum = 0;
            for (int k = 0; k < i; k++) {
                sum += L[i][k] * U[k][j];
            }
            U[i][j] = a[i][j] - sum;
        }

        for (int j = i + 1; j < n; j++) {
            double sum = 0;
            for (int k = 0; k < i; k++) {
                sum += L[j][k] * U[k][i];
            }
            L[j][i] = (a[j][i] - sum) / U[i][i];
        }
    }
}

int main() {
    int n;
    printf("Enter matrix size: ");
    scanf("%d", &n);

    double **a = (double **)malloc(n * sizeof(double *));
    double **L = (double **)malloc(n * sizeof(double *));
    double **U = (double **)malloc(n * sizeof(double *));
    
    for (int i = 0; i < n; i++) {
        a[i] = (double *)malloc(n * sizeof(double));
        L[i] = (double *)malloc(n * sizeof(double));
        U[i] = (double *)malloc(n * sizeof(double));
    }

    printf("Enter matrix row-wise:\n");
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            scanf("%lf", &a[i][j]);
        }
    }

    lu_decomposition(n, a, L, U);

    printf("L:\n");
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            printf("%.6f ", L[i][j]);
        }
        printf("\n");
    }

    printf("U:\n");
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            printf("%.6f ", U[i][j]);
        }
        printf("\n");
    }

    for (int i = 0; i < n; i++) {
        free(a[i]);
        free(L[i]);
        free(U[i]);
    }
    free(a);
    free(L);
    free(U);

    return 0;
}
